### PR TITLE
Add NEON implementation of vec_reduce_all()

### PR DESCRIPTION
**Pull request by Rohanjames1997**
As suggested in the [blog](https://dev-discuss.pytorch.org/t/torchinductor-update-5-cpu-backend-backend-performance-update-and-deep-dive-on-key-optimizations/1117#:~:text=It%20can%20be,sub%2Dclasses.), I subclassed the `VecISA` class and implemented a NEON version of the `vec_reduce_all()` function, to go along with the existing AVX2 and AVX512 versions. Any operation that calls `vec_reduce_all()` will also take the NEON path and benefit from its vectorization.

The `vec_reduce_all()` is invoked by Softmax and other operations like norms. Using the fast path results in 30% time savings for Softmax as compared to the previously taken slow path.

  | Slow path | Fast path (NEON intrinsics)
-- | -- | --
Softmax (100 passes, 1024 dimension) | 623.706ms | 452.011ms

### CODE REVIEW

File: aten/src/ATen/cpu/vec/functional_base.h [hunk 1]
```diff
@@ -76,6 +76,30 @@ struct VecReduceAllSIMD<float, Op> {
...
+    // 32-bit shuffle: [a7, a8, a5, a6,  a3, a4, a1, a2] -> [a8, a7, a6, a5, a4, a3, a2, a1]
+    v1_1 = vrev64q_f32(v.get_low());
+    v1_2 = vrev64q_f32(v.get_high());
+    v1 = {v1_1, v1_2};
+    v = vec_fun(v, v1);
+
+    return v.get_low()[0];
+  }
+};
+#endif // defined(CPU_CAPABILITY_NEON)
```
Comment (Rohanjames1997) [inline aten/src/ATen/cpu/vec/functional_base.h]: See [lines 38-56](https://github.com/Rohanjames1997/pytorch/pull/2/files#diff-0149700c2607377c8ea2741e6fcd9ee6adbba4ec49de6bb681375eb591932f60R38-R56) for reference AVX-2 code

File: torch/_inductor/codecache.py [hunk 2]
```diff
@@ -514,6 +515,17 @@ def __bool__(self):
 
             return True
 
+@dataclasses.dataclass
+class VecNEON(VecISA):
+    _bit_width = 256
```
Comment (AWSjswinney) [inline torch/_inductor/codecache.py]: Neon's bit width is 128. Do you need to use 256 in order to get it to use 256 in `VecReduceAllSIMD`? If so, I recommend writing a comment about that here.

**Review by AWSjswinney — CHANGES_REQUESTED** [review]

File: torch/_inductor/codecache.py (hunk 2 above)
Comment (Rohanjames1997) [reply torch/_inductor/codecache.py]: Yes, that is correct. I shall add a comment here.

In order to leverage the NEON functions implemented in [vec256_float_neon.h](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/cpu/vec/vec256/vec256_float_neon.h), the width needs to be 256.

**Comment by AWSjswinney** [comment]
I was thinking about this again. It really is just a reduce function. As long as the `vec_fun` is a communicable function, then we are doing just log2(8) operations (3) to get a result in the first lane. That mean we can definitely get rid of `v1_2`. I would also recommend trying the vertical add and vertical max intrinsics to see if that's faster for those cases. `svaddv_f32` and `vmaxv_f32`. Max has 4 cycle latency but add has 11 cycle latency for 32 bit floats. That's quite a bit slower, but it might still be worth it since we could avoid the `ext`, `rev64` instructions, and two uses of `vec_fun`.

For both you could do something like this: `vmaxv_f32(vmax_f32(a, b))`

**Comment by Rohanjames1997** [comment]
@AWSjswinney  you're right, we can definitely get rid of `v1_2`. I got rid of it after our last discussion but did not witness perf gains, most likely because the compiler was not executing the unnecessary instructions anyway like hypothesized. But I shall get red of them nonetheless.

Regarding `vec_fun` -  max and sum are just two instances of `vec_fun` that I have come across in my experiments. My guess is that there could be more reductions apart from these two. From a quick search, it seems that [these](https://github.com/pytorch/pytorch/blob/10cbc9a063d20116ded39fb54a59db9d17e53acf/torch/_inductor/codegen/cpp.py#L1526-L1532) are other possible reduction functions. Also note that `vec_reduce_all` is a function that is also called by generated code. 

So, one route I could take is to somehow check if `vec_fun` is indeed a max or a sum, and if so, execute the corresponding neon intrinsics that you suggested. I just feel that this would make it a bit harder to read & review for the PyTorch team, although I can bring it up with them once I have their attention. What do you think?

**Comment by yd2102** [comment]
> Regarding `vec_fun` - max and sum are just two instances of `vec_fun` that I have come across in my experiments. My guess is that there could be more reductions apart from these two. From a quick search, it seems that [these](https://github.com/pytorch/pytorch/blob/10cbc9a063d20116ded39fb54a59db9d17e53acf/torch/_inductor/codegen/cpp.py#L1526-L1532) are other possible reduction functions. Also note that `vec_reduce_all` is a function that is also called by generated code.

Yes we should treat that `vec_fun` as any possible kind of reduction operation. Based on offline discussion with @Rohanjames1997, `vec_reduce_all` isn't that expensive (correct me if I'm wrong) so even a 2x improvement in this function wouldn't make much difference. If we want to specialize for max or sum operations, it might be possible to leverage C++'s template specialization for this case.

**Comment by Rohanjames1997** [comment]
That is correct, `vec_reduce_all` isn't that expensive. As discussed, we can leave it as is, and possibly leverage C++'s templates in a later PR.
I've updated the function now to remove the unnecessary shuffles. I can raise the PR to upstream if no further comments.
